garage/tf/algos/ddpg.py

```diff
@@ -36,427 +29,241 @@ class DDPG(RLAlgorithm):
 
             with tf.name_scope("minimize_action_loss"):
-                actor_train_op = self.actor_optimizer(
-                    self.actor_lr, name="ActorOptimizer").minimize(
-                        action_loss, var_list=self.actor.trainable_vars)
+                policy_train_op = self.policy_optimizer(
+                    self.policy_lr, name="PolicyOptimizer").minimize(
+                        action_loss, var_list=self.policy.get_trainable_vars())
 
-            f_train_actor = tensor_utils.compile_function(
-                inputs=[obs], outputs=[actor_train_op, action_loss])
+            f_train_policy = tensor_utils.compile_function(
+                inputs=[obs], outputs=[policy_train_op, action_loss])
 
-            # Set up critic training function
-            qval = self.critic.get_qval_sym(obs, actions, name="q_value")
+            # Set up qf training function
+            qval = self.qf.get_qval_sym(obs, actions, name="q_value")
             with tf.name_scope("qval_loss"):
                 qval_loss = tf.reduce_mean(tf.squared_difference(y, qval))
-                if self.critic_weight_decay > 0.:
-                    critic_reg = tc.layers.apply_regularization(
-                        tc.layers.l2_regularizer(self.critic_weight_decay),
-                        weights_list=self.critic.regularizable_vars)
-                    qval_loss += critic_reg
-
-            with tf.name_scope("minimize_critic_loss"):
-                critic_train_op = self.critic_optimizer(
-                    self.critic_lr, name="CriticOptimizer").minimize(
-                        qval_loss, var_list=self.critic.trainable_vars)
-
-            f_train_critic = tensor_utils.compile_function(
+                if self.qf_weight_decay > 0.:
+                    qf_reg = tc.layers.apply_regularization(
+                        tc.layers.l2_regularizer(self.qf_weight_decay),
+                        weights_list=self.qf.get_regularizable_vars())
+                    qval_loss += qf_reg
+
+            with tf.name_scope("minimize_qf_loss"):
+                qf_train_op = self.qf_optimizer(
+                    self.qf_lr, name="QFunctionOptimizer").minimize(
+                        qval_loss, var_list=self.qf.get_trainable_vars())
+
+            f_train_qf = tensor_utils.compile_function(
                 inputs=[y, obs, actions],
-                outputs=[critic_train_op, qval_loss, qval])
+                outputs=[qf_train_op, qval_loss, qval])
 
-            self.f_train_actor = f_train_actor
-            self.f_train_critic = f_train_critic
+            self.f_train_policy = f_train_policy
+            self.f_train_qf = f_train_qf
             self.f_init_target = f_init_target
             self.f_update_target = f_update_target
-            self.replay_buffer = replay_buffer
-            self.target_critic = target_critic
-            self.target_actor = target_actor
 
-    def _learn(self):
+    @overrides
+    def train(self, sess=None):
```

Comment:
Are you sure that this can't go into `OffPolicyRLAlgorithm`?

Reply:
I don't think this can for now. The training process here is more customized to actor-critic algorithms, i.e. it calculates TD error. This doesn't fit DQN at least. We can generalize this later.